modyn/common/trigger_storage_cpp/trigger_storage_cpp.py

```diff
@@ -0,0 +1,433 @@
+        assert (retrieval_worker_id >= 0 and total_retrieval_workers >= 0) or (
+            retrieval_worker_id < 0 and total_retrieval_workers < 2
+        ), "Either both or none of the retrieval worker id must be negative and \
+            the total retrieval workers must be smaller than 2."
+        if retrieval_worker_id < 0 and total_retrieval_workers < 2:
+            return self._get_all_samples(pipeline_id, trigger_id, partition_id)
+        assert num_samples_trigger_partition > 0, "The number of samples per trigger must be positive."
+        return self._get_worker_samples(
+            pipeline_id,
+            trigger_id,
+            partition_id,
+            retrieval_worker_id,
+            total_retrieval_workers,
+            num_samples_trigger_partition,
+        )
+
+    def _get_worker_samples(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        partition_id: int,
+        retrieval_worker_id: int,
+        total_retrieval_workers: int,
+        num_samples_trigger_partition: int,
+    ) -> ArrayWrapper:
+        """
+        Return the trigger samples for the given pipeline id, trigger id and partition id that are assigned to the
+        retrieval worker.
+
+        :param pipeline_id: the id of the pipeline
+        :param trigger_id: the id of the trigger
+        :param partition_id: the id of the partition
+        :param retrieval_worker_id: the id of the retrieval worker
+        :param total_retrieval_workers: the total number of retrieval workers
+        :param num_samples_trigger_partition: the total number of samples per trigger and partition
+        :return: the trigger samples
+        """
+        start_index, worker_subset_size = get_partition_for_worker(
+            retrieval_worker_id, total_retrieval_workers, num_samples_trigger_partition
+        )
+
+        folder = ctypes.c_char_p(str(self.trigger_sample_directory).encode("utf-8"))
+        size = (ctypes.c_uint64 * 1)()
+        self.data_pointer.shape_val = size
+        pattern = ctypes.c_char_p(f"{pipeline_id}_{trigger_id}_{partition_id}_".encode("utf-8"))
+
+        data = self._get_worker_samples_impl(folder, size, pattern, start_index, worker_subset_size).reshape(-1)
+        result = ArrayWrapper(data, self._release_data_impl)
+        return result
+
+    def _get_all_samples(self, pipeline_id: int, trigger_id: int, partition_id: int) -> ArrayWrapper:
+        """
+        Return all the samples for the given pipeline id, trigger id and partition id.
+
+        :param pipeline_id: the id of the pipeline
+        :param trigger_id: the id of the trigger
+        :param partition_id: the id of the partition
+        :return: the trigger samples
+        """
+
+        folder = ctypes.c_char_p(str(self.trigger_sample_directory).encode("utf-8"))
+        size = (ctypes.c_uint64 * 1)()
+        self.data_pointer.shape_val = size
```

Comment:
Initially I thought that the library object was unique for every thread, as we discussed before, so it would've worked fine. Anyway, all this logic is gone now.